# Capstone Project - The Battle of Neighborhoods (Full Version)

## Challenge 1 - Business Understanding and Data Science Methodology

### Business Understanding

#### Business Scenarios

United States has been on top trending of paradise of new start up companies. There have been loads of immigrants to US with new business sections created and challenged. The trending generates more challenges for OPENING A BUSINESS (CAFE' AND RESTAURANT) in the Country. The City is diversed, competitive and captial of USA's Finance. The cosmopolitan in the city offers the City a lot of attractions to immigrants who wanna growth their Dream in the city.

Because of the aggressions of immigrants and high competition in New York City in terms of PLEASURING and CAFE' & RESTAURANT, an idea of creating a new business model that ENTER the market with a significant outstanding business in very short-time (6 months). There are 5 BOROUGHS in New York City which should be very potential to make the business idea happens. To firstly define strategic GO-TO-MARKET plan, we should analyse the city, 5 Boroughs and their Neighbourhoods to gain more insights from the city in terms of strategic location and population allocated. The ultimate of the project is to understand the market by strategical insights released based on Data Science in order to Business Owners make strategic decision for their CAFE' & RESTAURANT Business.

#### Business Problem Description

The perfect challenge is to ideate business model that provides insights to know which boroughs are ideal to open up a Cafe' & Restaurant based on the huge impact of population and neighbourhoods.

Big question is how we identify the perfect information of specifically idea BOROUGH and its NEIGHBOURHOODS in order to know the most impactful factor of business Enter. 

With Data Science, we will steps by steps to demonstrate the Cluster of New York City neighbourhoods. By giving the visualation of Cluster, we can recommend potential BOROUGH and strategic locations for ONWNERS for their decision making.

### Data Collection: 

City - New York City.

Datasets to understanding the City.

Source 1 - Neighbourhood Data: https://geo.nyu.edu/catalog/nyu-2451-34572

Soure 2 - New York Population: https://bit.ly/2StDkJJ (sources from Kaggle)

##### With multiple sourcesof data, we will access each of data to select target locations across different venues based on the accordance of filtering facitilities and factors. To make a JOIN to data of properties and pricing and other factors, we will explore 4 sources and plus with Foursquare API interface, we will be able to recommend affordable accommodation

### Measurables

1. Market Understanding of Strategic Locations
2. Ideal recommendation of accommodation in compared with Neighbourhood to Business Owners

## Methodology

1. Data Understanding Data - 
CRISP-DM combines the stages of Data Requirements, Data Collection, and Data Understanding. -> Data Preparation 
2. Data Preparation - Data Preprocessing - 
Transformed into a useable subset. Once a dataset is chosen, it must then be checked for questionable, missing, or ambiguous cases
3. Modeling - 
Select appropriate models, give meaningful insights, and hopefully new knowledge.

#### Let's start to do Data Science

# programming

### Import Libraries

In [294]:
import sys
!{sys.executable} -m pip install geocoder
!{sys.executable} -m pip install folium
import geocoder

from bs4 import BeautifulSoup
import numpy as np # library to handle data in a vectorized manner
import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
import json # library to handle JSON files
#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values
import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe
# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors
# import k-means from clustering stage
from sklearn.cluster import KMeans
from sklearn.datasets.samples_generator import make_blobs
#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library
from bs4 import BeautifulSoup
import lxml
print('Done')

Done


### Data Collection

In [295]:
# New York Dataset
!wget -q -O 'newyork_data.json' https://cocl.us/new_york_dataset
print('Data downloaded!')

Data downloaded!


In [296]:
with open('newyork_data.json') as json_data:
    newyork_data = json.load(json_data)

In [297]:
newyork_data

{'type': 'FeatureCollection',
 'totalFeatures': 306,
 'features': [{'type': 'Feature',
   'id': 'nyu_2451_34572.1',
   'geometry': {'type': 'Point',
    'coordinates': [-73.84720052054902, 40.89470517661]},
   'geometry_name': 'geom',
   'properties': {'name': 'Wakefield',
    'stacked': 1,
    'annoline1': 'Wakefield',
    'annoline2': None,
    'annoline3': None,
    'annoangle': 0.0,
    'borough': 'Bronx',
    'bbox': [-73.84720052054902,
     40.89470517661,
     -73.84720052054902,
     40.89470517661]}},
  {'type': 'Feature',
   'id': 'nyu_2451_34572.2',
   'geometry': {'type': 'Point',
    'coordinates': [-73.82993910812398, 40.87429419303012]},
   'geometry_name': 'geom',
   'properties': {'name': 'Co-op City',
    'stacked': 2,
    'annoline1': 'Co-op',
    'annoline2': 'City',
    'annoline3': None,
    'annoangle': 0.0,
    'borough': 'Bronx',
    'bbox': [-73.82993910812398,
     40.87429419303012,
     -73.82993910812398,
     40.87429419303012]}},
  {'type': 'Feature',
 

In [298]:
import urllib.request
url = "https://drive.google.com/file/d/1CL1NpJDEsrWJuPWITSioZjAq_rZ5fjR6/view?usp=sharing"
request = urllib.request.Request(url)
response = urllib.request.urlopen(request)
print (response.read().decode('utf-8'))

<!DOCTYPE html><html><head><meta name="google" content="notranslate"><meta http-equiv="X-UA-Compatible" content="IE=edge;"><style>@font-face{font-family:'Roboto';font-style:italic;font-weight:400;src:local('Roboto Italic'),local('Roboto-Italic'),url(//fonts.gstatic.com/s/roboto/v18/KFOkCnqEu92Fr1Mu51xIIzc.ttf)format('truetype');}@font-face{font-family:'Roboto';font-style:normal;font-weight:300;src:local('Roboto Light'),local('Roboto-Light'),url(//fonts.gstatic.com/s/roboto/v18/KFOlCnqEu92Fr1MmSU5fBBc9.ttf)format('truetype');}@font-face{font-family:'Roboto';font-style:normal;font-weight:400;src:local('Roboto Regular'),local('Roboto-Regular'),url(//fonts.gstatic.com/s/roboto/v18/KFOmCnqEu92Fr1Mu4mxP.ttf)format('truetype');}@font-face{font-family:'Roboto';font-style:normal;font-weight:700;src:local('Roboto Bold'),local('Roboto-Bold'),url(//fonts.gstatic.com/s/roboto/v18/KFOlCnqEu92Fr1MmWUlfBBc9.ttf)format('truetype');}</style><meta name="referrer" content="origin"><title>Newyork_Pop_Data.

In [299]:
neighborhoods_data = newyork_data["features"]
neighborhoods_data[0]

{'type': 'Feature',
 'id': 'nyu_2451_34572.1',
 'geometry': {'type': 'Point',
  'coordinates': [-73.84720052054902, 40.89470517661]},
 'geometry_name': 'geom',
 'properties': {'name': 'Wakefield',
  'stacked': 1,
  'annoline1': 'Wakefield',
  'annoline2': None,
  'annoline3': None,
  'annoangle': 0.0,
  'borough': 'Bronx',
  'bbox': [-73.84720052054902,
   40.89470517661,
   -73.84720052054902,
   40.89470517661]}}

In [300]:
# define the dataframe columns
column_names = ['Borough', 'Neighborhood', 'Latitude', 'Longitude'] 

# instantiate the dataframe
neighborhoods = pd.DataFrame(columns=column_names)

In [301]:
for data in neighborhoods_data:
    borough = neighborhood_name = data['properties']['borough'] 
    neighborhood_name = data['properties']['name']
        
    neighborhood_latlon = data['geometry']['coordinates']
    neighborhood_lat = neighborhood_latlon[1]
    neighborhood_lon = neighborhood_latlon[0]

    neighborhoods = neighborhoods.append({'Borough': borough,
                                          'Neighborhood': neighborhood_name,
                                          'Latitude': neighborhood_lat,
                                          'Longitude': neighborhood_lon}, ignore_index=True)

In [302]:
neighborhoods.head()

Borough Neighborhood   Latitude  Longitude
0   Bronx    Wakefield  40.894705 -73.847201
1   Bronx   Co-op City  40.874294 -73.829939
2   Bronx  Eastchester  40.887556 -73.827806
3   Bronx    Fieldston  40.895437 -73.905643
4   Bronx    Riverdale  40.890834 -73.912585

In [303]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(neighborhoods['Borough'].unique()),
        neighborhoods.shape[0]
    )
)

The dataframe has 5 boroughs and 306 neighborhoods.


### After counting the number of area, we figured out that in New York has 5 Boroughs and 306 Neighborhoods. To continue explore the potental Borough, we base on the total number of population on each borough, then pick the top one.

In [344]:
#Collect Population Data abd Read it in CSV
url = 'https://docs.google.com/spreadsheets/d/e/2PACX-1vTn0R4xaXtpuJHlpWm8mfULRPFixE5OAShoqE0RRHeBc4HaYrDhZeQI5ouEpa7IU215lZNzKCdR0s-L/pub?output=csv'
pop_data = pd.read_csv(url)
pop_data.head()

Borough Population
0   Bronx     91,497
1   Bronx     52,246
2   Bronx     79,762
3   Bronx    146,441
4   Bronx    128,200

In [345]:
# we set the both Borough in the same value of two tables. 
set(pop_data.Borough) - set(neighborhoods.Borough)

set()

In [346]:
# We add Population into the same columns to show a proper observation of Population on each Borough
Newyork_data_total = pd.merge(neighborhoods,pop_data, on='Borough')
Newyork_data_total.head()

Borough Neighborhood   Latitude  Longitude Population
0   Bronx    Wakefield  40.894705 -73.847201     91,497
1   Bronx    Wakefield  40.894705 -73.847201     52,246
2   Bronx    Wakefield  40.894705 -73.847201     79,762
3   Bronx    Wakefield  40.894705 -73.847201    146,441
4   Bronx    Wakefield  40.894705 -73.847201    128,200

In [307]:
Newyork_data_total.shape

(3687, 5)

In [308]:
# use the inline backend to generate the plots within the browser
%matplotlib inline 

import matplotlib as mpl
import matplotlib.pyplot as plt

mpl.style.use('ggplot') # optional: for ggplot-like style

# check for latest version of Matplotlib
print ('Matplotlib version: ', mpl.__version__) # >= 2.0.0

Matplotlib version:  3.0.2


In [309]:
borough_pop_sum = Newyork_data_total.groupby('Borough', as_index=False).agg({"Population": "sum"})
borough_pop_sum.head()

Borough                                         Population
0          Bronx  91,49752,24679,762146,441128,20083,268139,2861...
1       Brooklyn  173,08399,617152,985112,634182,896104,709126,2...
2      Manhattan  60,97890,016163,277103,24551,673142,745209,084...
3         Queens  191,105113,200171,576172,598169,190113,257247,...
4  Staten Island  175,756132,003160,209175,756132,003160,209175,...

In [310]:
df_top5 = borough_pop_sum.sort_values(by=['Borough', 'Population'], ascending=False)

df_top5.head()

Borough                                         Population
4  Staten Island  175,756132,003160,209175,756132,003160,209175,...
3         Queens  191,105113,200171,576172,598169,190113,257247,...
2      Manhattan  60,97890,016163,277103,24551,673142,745209,084...
1       Brooklyn  173,08399,617152,985112,634182,896104,709126,2...
0          Bronx  91,49752,24679,762146,441128,20083,268139,2861...

#### we have seen the highest Population belongs to Borough "Staten Island" -> This is potential Borough to go analysis ahead

### discovering the Staten Island and visualise it to see strategic locations of the borough and its neighborhoods

In [347]:
address = 'New York City, NY'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of New York City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of New York City are 40.7127281, -74.0060152.


In [348]:
# create map of New York using latitude and longitude values
map_newyork = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(neighborhoods['Latitude'], neighborhoods['Longitude'], neighborhoods['Borough'], neighborhoods['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_newyork)  
    
map_newyork

In [349]:
# We explore exactly the Staten Island and its long, lat
StI_data = neighborhoods[neighborhoods['Borough'] == 'Staten Island'].reset_index(drop=True)
StI_data.head()

Borough   Neighborhood   Latitude  Longitude
0  Staten Island     St. George  40.644982 -74.079353
1  Staten Island   New Brighton  40.640615 -74.087017
2  Staten Island      Stapleton  40.626928 -74.077902
3  Staten Island       Rosebank  40.615305 -74.069805
4  Staten Island  West Brighton  40.631879 -74.107182

In [350]:
address = 'Staten Island, NY'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Staten Island are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Staten Island are 40.5834557, -74.1496048.


In [315]:
# create map of Manhattan using latitude and longitude values
map_StI = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(StI_data['Latitude'], StI_data['Longitude'], StI_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_StI)  
    
map_StI

#### Now we discover the Foursquare API to see the market an nearby areas of Staten Island for Cafe' & Restaurant and other facilities

In [352]:
CLIENT_ID = 'IGFAY3CHU0CGOTDGFOYL2M4TAFFM2RA0DU1E0JBHYNVF4PLH' # your Foursquare ID
CLIENT_SECRET = 'WNF4IBUV3DFOLHYBLNJRTDXOAWIXL3T3OZHUBH1DIOAELQEY' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: IGFAY3CHU0CGOTDGFOYL2M4TAFFM2RA0DU1E0JBHYNVF4PLH
CLIENT_SECRET:WNF4IBUV3DFOLHYBLNJRTDXOAWIXL3T3OZHUBH1DIOAELQEY


##### we deep dive one neighborhood of Staten Island to see what else it has

In [355]:
StI_data.loc[0, 'Neighborhood']

'St. George'

In [356]:
neighborhood_latitude = StI_data.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = StI_data.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = StI_data.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of St. George are 40.6449815710044, -74.07935312512797.


In [357]:
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius

# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=IGFAY3CHU0CGOTDGFOYL2M4TAFFM2RA0DU1E0JBHYNVF4PLH&client_secret=WNF4IBUV3DFOLHYBLNJRTDXOAWIXL3T3OZHUBH1DIOAELQEY&v=20180605&ll=40.6449815710044,-74.07935312512797&radius=500&limit=100'

In [358]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5e3e88601e152c001b4d267d'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': '$-$$$$', 'key': 'price'}]},
  'headerLocation': 'Current map view',
  'headerFullLocation': 'Current map view',
  'headerLocationGranularity': 'unknown',
  'totalResults': 35,
  'suggestedBounds': {'ne': {'lat': 40.6494815755044,
    'lng': -74.07343346476772},
   'sw': {'lat': 40.6404815665044, 'lng': -74.08527278548821}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4bb6924c46d4a5938e7ac6c0',
       'name': 'A&S Pizzeria',
       'location': {'address': '87 Stuyvesant Pl',
        'crossStreet': 'Wall st',
        'lat': 40.64393953223924,
        'lng': -74.0776259226109,
        'labeledLatLngs': [{'label': 'display

In [359]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [360]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name           categories        lat  \
0                         A&S Pizzeria          Pizza Place  40.643940   
1                                 Beso     Tapas Restaurant  40.643306   
2        Richmond County Bank Ballpark     Baseball Stadium  40.645056   
3  Staten Island September 11 Memorial  Monument / Landmark  40.646767   
4                   Nike Factory Store  Sporting Goods Shop  40.645753   

         lng  
0 -74.077626  
1 -74.076508  
2 -74.076864  
3 -74.076510  
4 -74.077702

#### at St. George area, there are including categories: Pizza, Tapas, Stadium, Landmark, Sport Goods

In [362]:
print('{} venues were returned by Foursquare in St.George.'.format(nearby_venues.shape[0]))

35 venues were returned by Foursquare in St.George.


### Now it's time to discover the Nearby Venues of Staten Island to see the total number of venues, then map them together and counting all categories included all 

In [363]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [364]:
StI_venues = getNearbyVenues(names=StI_data['Neighborhood'],
                                   latitudes=StI_data['Latitude'],
                                   longitudes=StI_data['Longitude']
                                  )

St. George
New Brighton
Stapleton
Rosebank
West Brighton
Grymes Hill
Todt Hill
South Beach
Port Richmond
Mariner's Harbor
Port Ivory
Castleton Corners
New Springville
Travis
New Dorp
Oakwood
Great Kills
Eltingville
Annadale
Woodrow
Tottenville
Tompkinsville
Silver Lake
Sunnyside
Park Hill
Westerleigh
Graniteville
Arlington
Arrochar
Grasmere
Old Town
Dongan Hills
Midland Beach
Grant City
New Dorp Beach
Bay Terrace
Huguenot
Pleasant Plains
Butler Manor
Charleston
Rossville
Arden Heights
Greenridge
Heartland Village
Chelsea
Bloomfield
Bulls Head
Richmond Town
Shore Acres
Clifton
Concord
Emerson Hill
Randall Manor
Howland Hook
Elm Park
Manor Heights
Willowbrook
Sandy Ground
Egbertville
Prince's Bay
Lighthouse Hill
Richmond Valley
Fox Hills


In [367]:
print(StI_venues.shape)
StI_venues.head()

(829, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0   St. George              40.644982              -74.079353   
1   St. George              40.644982              -74.079353   
2   St. George              40.644982              -74.079353   
3   St. George              40.644982              -74.079353   
4   St. George              40.644982              -74.079353   

                                 Venue  Venue Latitude  Venue Longitude  \
0                         A&S Pizzeria       40.643940       -74.077626   
1                                 Beso       40.643306       -74.076508   
2        Richmond County Bank Ballpark       40.645056       -74.076864   
3  Staten Island September 11 Memorial       40.646767       -74.076510   
4                   Nike Factory Store       40.645753       -74.077702   

        Venue Category  
0          Pizza Place  
1     Tapas Restaurant  
2     Baseball Stadium  
3  Monument / Landmark  
4  Sporting Goods Shop

#### First, we now know there are 829 Venues nearby Staten Island

In [368]:
StI_venues.groupby('Neighborhood').count()

Neighborhood Latitude  Neighborhood Longitude  Venue  \
Neighborhood                                                              
Annadale                              10                      10     10   
Arden Heights                          4                       4      4   
Arlington                              5                       5      5   
Arrochar                              18                      18     18   
Bay Terrace                            9                       9      9   
Bloomfield                             5                       5      5   
Bulls Head                            44                      44     44   
Butler Manor                           5                       5      5   
Castleton Corners                     15                      15     15   
Charleston                            28                      28     28   
Chelsea                                7                       7      7   
Clifton                               20                      20     20   
Concord                                9                       9      9   
Dongan Hills                          25                      25     25   
Egbertville                            5                       5      5   
Elm Park                               9                       9      9   
Eltingville                           43                      43     43   
Emerson Hill                           3                       3      3   
Fox Hills                              5                       5      5   
Graniteville                           4                       4      4   
Grant City                            23                      23     23   
Grasmere                              22                      22     22   
Great Kills                           20                      20     20   
Greenridge                             8                       8      8   
Grymes Hill                            1                       1      1   
Heartland Village                     11                      11     11   
Huguenot                               8                       8      8   
Lighthouse Hill                        6                       6      6   
Manor Heights                         11                      11     11   
Mariner's Harbor                       5                       5      5   
Midland Beach                         12                      12     12   
New Brighton                          10                      10     10   
New Dorp                              26                      26     26   
New Dorp Beach                        12                      12     12   
New Springville                       27                      27     27   
Oakwood                                2                       2      2   
Old Town                              19                      19     19   
Park Hill                              7                       7      7   
Pleasant Plains                       16                      16     16   
Port Richmond                          6                       6      6   
Prince's Bay                          10                      10     10   
Randall Manor                          3                       3      3   
Richmond Town                          5                       5      5   
Richmond Valley                        9                       9      9   
Rosebank                              24                      24     24   
Rossville                             17                      17     17   
Sandy Ground                           9                       9      9   
Shore Acres                           25                      25     25   
Silver Lake                            3                       3      3   
South Beach                            8                       8      8   
St. George                            35                      35     35   
Stapleton                             31    

In [369]:
print('There are {} uniques categories.'.format(len(StI_venues['Venue Category'].unique())))

There are 183 uniques categories.


### There 183 Categories in Venues nearby, offering the large choices for Citizen, however, which Categories are the most selected to open?

In [370]:
# one hot encoding
StI_onehot = pd.get_dummies(StI_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
StI_onehot['Neighborhood'] = StI_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [StI_onehot.columns[-1]] + list(StI_onehot.columns[:-1])
StI_onehot = StI_onehot[fixed_columns]

StI_onehot.head()

Yoga Studio  American Restaurant  Arcade  Art Gallery  Art Museum  \
0            0                    0       0            0           0   
1            0                    0       0            0           0   
2            0                    0       0            0           0   
3            0                    0       0            0           0   
4            0                    0       0            0           0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  BBQ Joint  \
0                    0                 0                   0          0   
1                    0                 0                   0          0   
2                    0                 0                   0          0   
3                    0                 0                   0          0   
4                    0                 0                   0          0   

   Bagel Shop  Bakery  Bank  Bar  Baseball Field  Baseball Stadium  \
0           0       0     0    0               0                 0   
1           0       0     0    0               0                 0   
2           0       0     0    0               0                 1   
3           0       0     0    0               0                 0   
4           0       0     0    0               0                 0   

   Basketball Court  Beach  Beer Bar  Big Box Store  Board Shop  \
0                 0      0         0              0           0   
1                 0      0         0              0           0   
2                 0      0         0              0           0   
3                 0      0         0              0           0   
4                 0      0         0              0           0   

   Boat or Ferry  Bookstore  Bowling Alley  Breakfast Spot  Brewery  \
0              0          0              0               0        0   
1              0          0              0               0        0   
2              0          0              0               0        0   
3              0          0              0               0        0   
4              0          0              0               0        0   

   Burger Joint  Bus Stop  Business Service  Cafeteria  Café  \
0             0         0                 0          0     0   
1             0         0                 0          0     0   
2             0         0                 0          0     0   
3             0         0                 0          0     0   
4             0         0                 0          0     0   

   Cajun / Creole Restaurant  Campground  Caribbean Restaurant  \
0                          0           0                     0   
1                          0           0                     0   
2                          0           0                     0   
3                          0           0                     0   
4                          0           0                     0   

   Check Cashing Service  Child Care Service  Chinese Restaurant  \
0                      0                   0                   0   
1                      0                   0                   0   
2                      0                   0                   0   
3                      0                   0                   0   
4                      0                   0                   0   

   Clothing Store  Cocktail Bar  Coffee Shop  Comedy Club  Concert Hall  \
0               0             0            0            0             0   
1               0             0            0            0             0   
2               0             0            0            0             0   
3               0             0            0            0             0   
4               0             0            0            0             0   

   Construction & Landscaping  Convenience Store  Cosmetics Shop  \
0                           0                  0               0   
1                           0                  0               0   
2                     

In [371]:
StI_onehot.shape

(829, 183)

In [372]:
StI_grouped = StI_onehot.groupby('Neighborhood').mean().reset_index()
StI_grouped

Neighborhood  Yoga Studio  American Restaurant    Arcade  \
0            Annadale         0.00             0.000000  0.000000   
1       Arden Heights         0.00             0.000000  0.000000   
2           Arlington         0.00             0.000000  0.000000   
3            Arrochar         0.00             0.000000  0.000000   
4         Bay Terrace         0.00             0.000000  0.000000   
5          Bloomfield         0.00             0.000000  0.000000   
6          Bulls Head         0.00             0.022727  0.000000   
7        Butler Manor         0.00             0.000000  0.000000   
8   Castleton Corners         0.00             0.000000  0.000000   
9          Charleston         0.00             0.035714  0.000000   
10            Chelsea         0.00             0.000000  0.000000   
11            Clifton         0.00             0.000000  0.000000   
12            Concord         0.00             0.000000  0.000000   
13       Dongan Hills         0.00             0.000000  0.000000   
14        Egbertville         0.00             0.000000  0.000000   
15           Elm Park         0.00             0.111111  0.000000   
16        Eltingville         0.00             0.000000  0.000000   
17       Emerson Hill         0.00             0.000000  0.000000   
18          Fox Hills         0.00             0.000000  0.000000   
19       Graniteville         0.00             0.000000  0.000000   
20         Grant City         0.00             0.000000  0.000000   
21           Grasmere         0.00             0.000000  0.000000   
22        Great Kills         0.00             0.000000  0.000000   
23         Greenridge         0.00             0.000000  0.000000   
24        Grymes Hill         0.00             0.000000  0.000000   
25  Heartland Village         0.00             0.000000  0.000000   
26           Huguenot         0.00             0.000000  0.000000   
27    Lighthouse Hill         0.00             0.000000  0.000000   
28      Manor Heights         0.00             0.090909  0.000000   
29   Mariner's Harbor         0.00             0.000000  0.000000   
30      Midland Beach         0.00             0.000000  0.000000   
31       New Brighton         0.00             0.000000  0.000000   
32           New Dorp         0.00             0.000000  0.000000   
33     New Dorp Beach         0.00             0.000000  0.000000   
34    New Springville         0.00             0.000000  0.000000   
35            Oakwood         0.00             0.000000  0.000000   
36           Old Town         0.00             0.052632  0.000000   
37          Park Hill         0.00             0.000000  0.000000   
38    Pleasant Plains         0.00             0.000000  0.000000   
39      Port Richmond         0.00             0.000000  0.000000   
40       Prince's Bay         0.00             0.000000  0.000000   
41      Randall Manor         0.00             0.000000  0.000000   
42      Richmond Town         0.00             0.000000  0.000000   
43    Richmond Valley         0.00             0.000000  0.000000   
44           Rosebank         0.00             0.000000  0.000000   
45          Rossville         0.00             0.058824  0.000000   
46       Sandy Ground         0.00             0.000000  0.000000   
47        Shore Acres         0.04             0.000000  0.000000   
48        Silver Lake         0.00             0.333333  0.000000   
49        South Beach         0.00             0.000000  0.000000   
50         St. George         0.00             0.028571  0.000000   
51          Stapleton         0.00             0.032258  0.000000   
52          Sunnyside         0.00             0.125000  0.000000   
53          Todt Hill         0.00             0.000000  0.000000   
54      Tompkinsville         0.00             0.000000  0.000000   
55        Tottenville         0.00             0.000000  0.000000   
56             Travis         0.00             0.000000  0.000000   
57     

In [373]:
StI_grouped.shape

(61, 183)

In [374]:
num_top_venues = 5

for hood in StI_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = StI_grouped[StI_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Annadale----
              venue  freq
0            Bakery   0.2
1        Restaurant   0.1
2             Diner   0.1
3        Sports Bar   0.1
4  Sushi Restaurant   0.1


----Arden Heights----
         venue  freq
0  Pizza Place  0.25
1     Bus Stop  0.25
2  Coffee Shop  0.25
3     Pharmacy  0.25
4  Yoga Studio  0.00


----Arlington----
           venue  freq
0       Bus Stop   0.2
1  Deli / Bodega   0.2
2  Boat or Ferry   0.2
3  Grocery Store   0.2
4    Coffee Shop   0.2


----Arrochar----
                venue  freq
0  Italian Restaurant  0.11
1            Bus Stop  0.11
2       Deli / Bodega  0.11
3               Hotel  0.06
4         Supermarket  0.06


----Bay Terrace----
                venue  freq
0         Supermarket  0.22
1  Italian Restaurant  0.11
2    Insurance Office  0.11
3          Donut Shop  0.11
4       Train Station  0.11


----Bloomfield----
               venue  freq
0  Recreation Center   0.2
1               Park   0.2
2           Bus Stop   0.2
3         The

In [375]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [376]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = StI_grouped['Neighborhood']

for ind in np.arange(StI_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(StI_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0       Annadale                Bakery          Dance Studio   
1  Arden Heights              Pharmacy              Bus Stop   
2      Arlington         Deli / Bodega              Bus Stop   
3       Arrochar              Bus Stop         Deli / Bodega   
4    Bay Terrace           Supermarket      Insurance Office   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0           Pizza Place      Sushi Restaurant                  Park   
1           Coffee Shop           Pizza Place           Event Space   
2         Boat or Ferry           Coffee Shop         Grocery Store   
3    Italian Restaurant            Food Truck     Polish Restaurant   
4    Italian Restaurant         Train Station    Salon / Barbershop   

   6th Most Common Venue 7th Most Common Venue      8th Most Common Venue  \
0             Restaurant            Sports Bar                      Diner   
1      Food & Drink Shop                  Food                Flower Shop   
2   Fast Food Restaurant            Food Truck          Food & Drink Shop   
3  Outdoors & Recreation           Supermarket  Middle Eastern Restaurant   
4             Donut Shop      Sushi Restaurant             Shipping Store   

  9th Most Common Venue 10th Most Common Venue  
0         Train Station      Food & Drink Shop  
1     Fish & Chips Shop    Filipino Restaurant  
2                  Food            Flower Shop  
3        Sandwich Place             Bagel Shop  
4           Wings Joint            Event Space

In [377]:
# set number of clusters
kclusters = 5

StI_grouped_clustering = StI_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(StI_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 3, 3, 0, 0, 3, 0, 0, 0, 0], dtype=int32)

In [337]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

StI_merged = StI_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
StI_merged = StI_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

StI_merged.head() # check the last columns!

Borough   Neighborhood   Latitude  Longitude  Cluster Labels  \
0  Staten Island     St. George  40.644982 -74.079353             0.0   
1  Staten Island   New Brighton  40.640615 -74.087017             3.0   
2  Staten Island      Stapleton  40.626928 -74.077902             0.0   
3  Staten Island       Rosebank  40.615305 -74.069805             0.0   
4  Staten Island  West Brighton  40.631879 -74.107182             0.0   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0        Clothing Store                   Bar   Sporting Goods Shop   
1              Bus Stop                  Park         Bowling Alley   
2        Discount Store           Pizza Place    Mexican Restaurant   
3           Pizza Place         Grocery Store    Italian Restaurant   
4           Coffee Shop        Sandwich Place   American Restaurant   

  4th Most Common Venue 5th Most Common Venue  \
0    Italian Restaurant                 Plaza   
1        Discount Store          Home Service   
2                  Bank        Sandwich Place   
3              Pharmacy        Cosmetics Shop   
4    Italian Restaurant           Music Store   

                      6th Most Common Venue        7th Most Common Venue  \
0                            Scenic Lookout                 Burger Joint   
1                                Playground         Fast Food Restaurant   
2  Residential Building (Apartment / Condo)                   Skate Park   
3                            Discount Store  Eastern European Restaurant   
4                                      Bank                          Bar   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0        Farmers Market           Coffee Shop             Donut Shop  
1            Food Truck     Food & Drink Shop                   Food  
2       Motorcycle Shop  Fast Food Restaurant             Restaurant  
3            Restaurant        Sandwich Place    Filipino Restaurant  
4           Wings Joint          Burger Joint     Mexican Restaurant

#### Now we see Clothing, Bus, Pizza, Coffee are the 1st Most Common Venue.  "Restaurant" is the most service across the Common Venues.

In [378]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(StI_merged['Latitude'], StI_merged['Longitude'], StI_merged['Neighborhood'], StI_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow,
        fill=True,
        fill_color=rainbow,
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## Examine Cluster

In [379]:
#Cluster 1
StI_merged.loc[StI_merged['Cluster Labels'] == 0, StI_merged.columns[[1] + list(range(5, StI_merged.shape[1]))]]

Neighborhood       1st Most Common Venue  \
0          St. George              Clothing Store   
2           Stapleton              Discount Store   
3            Rosebank                 Pizza Place   
4       West Brighton                 Coffee Shop   
7         South Beach               Deli / Bodega   
9    Mariner's Harbor          Italian Restaurant   
11  Castleton Corners                 Pizza Place   
12    New Springville           Mobile Phone Shop   
13             Travis                       Hotel   
14           New Dorp                 Pizza Place   
15            Oakwood                         Bar   
16        Great Kills                         Bar   
17        Eltingville                 Pizza Place   
18           Annadale                      Bakery   
19            Woodrow                    Pharmacy   
20        Tottenville               Deli / Bodega   
21      Tompkinsville                     Brewery   
23          Sunnyside                     Theater   
25        Westerleigh                      Arcade   
28           Arrochar                    Bus Stop   
29           Grasmere                    Bus Stop   
30           Old Town          Italian Restaurant   
31       Dongan Hills                 Pizza Place   
32      Midland Beach                       Beach   
33         Grant City                 Pizza Place   
34     New Dorp Beach               Deli / Bodega   
35        Bay Terrace                 Supermarket   
36           Huguenot              Ice Cream Shop   
37    Pleasant Plains                 Pizza Place   
38       Butler Manor                        Pool   
39         Charleston              Cosmetics Shop   
40          Rossville                 Pizza Place   
42         Greenridge  Construction & Landscaping   
43  Heartland Village                 Coffee Shop   
44            Chelsea          Spanish Restaurant   
46         Bulls Head                    Bus Stop   
47      Richmond Town          Italian Restaurant   
48        Shore Acres                    Bus Stop   
49            Clifton               Train Station   
50            Concord              Cosmetics Shop   
51       Emerson Hill  Construction & Landscaping   
54           Elm Park          Italian Restaurant   
55      Manor Heights               Deli / Bodega   
57       Sandy Ground                Intersection   
58        Egbertville          Italian Restaurant   
59       Prince's Bay                 Pizza Place   
60    Lighthouse Hill               Moving Target   
61    Richmond Valley               Deli / Bodega   

          2nd Most Common Venue       3rd Most Common Venue  \
0                           Bar         Sporting Goods Shop   
2                   Pizza Place          Mexican Restaurant   
3                 Grocery Store          Italian Restaurant   
4                Sandwich Place         American Restaurant   
7                         Beach                        Pier   
9                 Deli / Bodega                    Bus Stop   
11                         Bank              Sandwich Place   
12      Health & Beauty Service                 Coffee Shop   
13                Deli / Bodega               Bowling Alley   
14           Italian Restaurant                Dessert Shop   
15                       Lawyer                 Wings Joint   
16                  Pizza Place          Italian Restaurant   
17             Sushi Restaurant          Italian Restaurant   
18                 Dance Studio                 Pizza Place   
19                  Coffee Shop          Chinese Restaurant   
20           Italian Restaurant                    Bus Stop   
21                Deli / Bodega                        Café   
23          American Restaurant                         Spa   
25                  Art Gallery           Convenience Store   
28                Deli / Bodega          Italian Restaurant   
29                   Bagel Shop               Grocery Store   
30                  Gas S

In [380]:
#Cluster 2
StI_merged.loc[StI_merged['Cluster Labels'] == 1, StI_merged.columns[[1] + list(range(5, StI_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
6    Todt Hill                  Park           Wings Joint   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
6  Furniture / Home Store            Food Truck     Food & Drink Shop   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
6                  Food           Flower Shop     Fish & Chips Shop   

  9th Most Common Venue 10th Most Common Venue  
6   Filipino Restaurant   Fast Food Restaurant

In [381]:
#Cluster 3
StI_merged.loc[StI_merged['Cluster Labels'] == 2, StI_merged.columns[[1] + list(range(5, StI_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
5  Grymes Hill               Dog Run           Wings Joint   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
5        Farmers Market     French Restaurant            Food Truck   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
5     Food & Drink Shop                  Food           Flower Shop   

  9th Most Common Venue 10th Most Common Venue  
5     Fish & Chips Shop    Filipino Restaurant

In [382]:
# Cluster 4
StI_merged.loc[StI_merged['Cluster Labels'] == 3, StI_merged.columns[[1] + list(range(5, StI_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
1    New Brighton              Bus Stop                  Park   
8   Port Richmond              Bus Stop   Rental Car Location   
24      Park Hill              Bus Stop  Gym / Fitness Center   
26   Graniteville              Bus Stop         Boat or Ferry   
27      Arlington         Deli / Bodega              Bus Stop   
41  Arden Heights              Pharmacy              Bus Stop   
45     Bloomfield     Recreation Center              Bus Stop   
52  Randall Manor              Bus Stop           Wings Joint   
56    Willowbrook              Bus Stop    Chinese Restaurant   
62      Fox Hills              Bus Stop            Playground   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
1          Bowling Alley        Discount Store          Home Service   
8             Donut Shop           Pizza Place      Basketball Court   
24                 Hotel    Athletics & Sports                  Park   
26        Sandwich Place         Grocery Store           Wings Joint   
27         Boat or Ferry           Coffee Shop         Grocery Store   
41           Coffee Shop           Pizza Place           Event Space   
45        Discount Store                  Park            Theme Park   
52        Farmers Market     French Restaurant            Food Truck   
56          Intersection           Pizza Place                   Spa   
62        Sandwich Place          Cocktail Bar           Wings Joint   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
1             Playground  Fast Food Restaurant            Food Truck   
8            Wings Joint  Fast Food Restaurant            Food Truck   
24           Coffee Shop           Wings Joint     Food & Drink Shop   
26        Farmers Market     Food & Drink Shop                  Food   
27  Fast Food Restaurant            Food Truck     Food & Drink Shop   
41     Food & Drink Shop                  Food           Flower Shop   
45           Event Space     Food & Drink Shop                  Food   
52     Food & Drink Shop                  Food           Flower Shop   
56         Deli / Bodega          Home Service                 Hotel   
62           Event Space     Food & Drink Shop                  Food   

   9th Most Common Venue 10th Most Common Venue  
1      Food & Drink Shop                   Food  
8      Food & Drink Shop                   Food  
24                  Food            Flower Shop  
26           Flower Shop      Fish & Chips Shop  
27                  Food            Flower Shop  
41     Fish & Chips Shop    Filipino Restaurant  
45           Flower Shop      Fish & Chips Shop  
52     Fish & Chips Shop    Filipino Restaurant  
56         Hot Dog Joint            Flower Shop  
62           Flower Shop      Fish & Chips Shop

In [383]:
# Cluster 5
StI_merged.loc[StI_merged['Cluster Labels'] == 4, StI_merged.columns[[1] + list(range(5, StI_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
22  Silver Lake   American Restaurant          Burger Joint   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
22           Golf Course           Wings Joint  Fast Food Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
22     French Restaurant            Food Truck     Food & Drink Shop   

   9th Most Common Venue 10th Most Common Venue  
22                  Food            Flower Shop